### PR TITLE
support subscription diagnosticsetting

### PR DESCRIPTION
**Pull request by VeryEarly**
## Checklist

- [ ] I have read the [_Submitting Changes_](../blob/master/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](../blob/master/CONTRIBUTING.md)
- [ ] The title of the PR is clear and informative
- [ ] The appropriate `ChangeLog.md` file(s) has been updated:
    - For any service, the `ChangeLog.md` file can be found at `src/{{SERVICE}}/{{SERVICE}}/ChangeLog.md`
    - A snippet outlining the change(s) made in the PR should be written under the `## Upcoming Release` header -- no new version header should be added
- [ ] The PR does not introduce [breaking changes](../blob/master/documentation/breaking-changes/breaking-changes-definition.md)
- [ ] If applicable, the changes made in the PR have proper test coverage
- [ ] For public API changes to cmdlets:
    - [ ] a cmdlet design review was approved for the changes in [this repository](https://github.com/Azure/azure-powershell-cmdlet-review-pr) (_Microsoft internal only_)
        - {Please put the link here}
    - [ ] the markdown help files have been regenerated using the commands listed [here](../blob/master/documentation/development-docs/help-generation.md#updating-all-markdown-files-in-a-module)

### CODE REVIEW

File: src/Monitor/Monitor/ChangeLog.md [hunk 1]
```diff
@@ -19,6 +19,12 @@
 -->
 
 ## Upcoming Release
+* Added cmdlet to get diagnostic setting categories for subscription
+    - `Get-AzSubscriptionDiagnosticSettingCategory`
+* Supported subscription diagnostic setting operations with new parameter: SubscrrptionId
```
Comment (BethanyZhou) [inline src/Monitor/Monitor/ChangeLog.md:24]:
```suggestion
* Supported subscription diagnostic setting operations with new parameter: SubscriptionId
```

File: src/Monitor/Monitor/Diagnostics/DiagnosticSettingCommandBase.cs [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+        [Parameter(ParameterSetName = SubscriptionIdParameterSet, Mandatory = true, ValueFromPipelineByPropertyName = true, HelpMessage = "The subscription id")]
+        [ValidateNotNullOrEmpty]
+        public string SubscriptionId { get; set; }
+
+        protected virtual void Validate()
+        {
+            if (this.IsParameterBound(c => c.SubscriptionId) && !Utilities.IsGuid(this.SubscriptionId))
+            {
+                throw new PSArgumentException("Please specify a valid subscription Id");
```
Comment (BethanyZhou) [inline src/Monitor/Monitor/Diagnostics/DiagnosticSettingCommandBase.cs:38]: It's encouraged to use AzPSArgumentException

File: src/Monitor/Monitor/Diagnostics/DiagnosticSettingCommandBase.cs [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+            }
+
+            if (this.IsParameterBound(c => c.ResourceId) && Utilities.IsGuid(this.ResourceId))
+            {
+                throw new PSArgumentException("Please use parameter 'SubscriptionId' instead of 'ResourceId'");
```
Comment (BethanyZhou) [inline src/Monitor/Monitor/Diagnostics/DiagnosticSettingCommandBase.cs:43]: Why do not you say "Please use correct ResourceId"?

Comment (VeryEarly) [reply src/Monitor/Monitor/Diagnostics/DiagnosticSettingCommandBase.cs]: this if block executed only when user use ResourceId and input is actually a subscriptionId.

Comment (VeryEarly) [reply src/Monitor/Monitor/Diagnostics/DiagnosticSettingCommandBase.cs]: > Why do not you say "Please use correct ResourceId"?

this if block executed only when user use parameter: "ResourceId" but input is actually a subscriptionId.

File: src/Monitor/Monitor/Utilities.cs [hunk 4]
```diff
@@ -84,5 +84,11 @@ public static string ReadFileContent(string path)
             using (TextReader reader = new StreamReader(path))
                 return reader.ReadToEnd();
         }
+
+        public static bool IsGuid(string str)
+        {
+            Guid g;
+            return Guid.TryParse(str, out g);
```
Comment (isra-fel) [inline src/Monitor/Monitor/Utilities.cs:91]: Syntax sugar:
```suggestion
            return Guid.TryParse(str, out _);
```
see [Discards](https://docs.microsoft.com/en-us/dotnet/csharp/discards)

File: src/Monitor/Monitor/help/Set-AzDiagnosticSetting.md [hunk 5]
```diff
@@ -125,6 +125,21 @@ PS C:\>Get-AzDiagnosticSetting -ResourceId "Resource01" | Set-AzDiagnosticSettin
...
+```powershell
+	$list = @()
+	Get-AzSubscriptionDiagnosticSettingCategory | ForEach-Object {
+		$list += (New-AzDiagnosticDetailSetting -Log -Category $_.Name -Enabled)
+	}
+	$DiagnosticSettingName = 'please use your setting name here'
+	$SubscriptionId = 'please use your subscription Id here'
+	$WorkspaceId = 'please use your workspace Id here'
+	$setting = New-AzDiagnosticSetting -Name $DiagnosticSettingName -SubscriptionId $SubscriptionId -WorkspaceId $WorkspaceId -Setting $list
+	Set-AzDiagnosticSetting -InputObject $setting
```
Comment (isra-fel) [inline src/Monitor/Monitor/help/Set-AzDiagnosticSetting.md:138]: suggest removing leading spaces

File: src/Monitor/Monitor/help/Get-AzDiagnosticSetting.md [hunk 6]
```diff
@@ -13,8 +13,15 @@ Gets the logged categories and time grains.
 
 ## SYNTAX
 
+### ResourceIdParameterSet
 ```
-Get-AzDiagnosticSetting [-ResourceId] <String> [-Name <String>] [-DefaultProfile <IAzureContextContainer>]
+Get-AzDiagnosticSetting [-Name <String>] -ResourceId <String> [-DefaultProfile <IAzureContextContainer>]
```
Comment (isra-fel) [inline src/Monitor/Monitor/help/Get-AzDiagnosticSetting.md:18]: breaking change: positional -> non-positional

**Review by isra-fel — APPROVED** [review]

**Pull request merged**